### ISSUE TITLE
Numpy Dependency (1.24.2) does not work with Python3.12

### ISSUE THREAD

**Issue by AndreVale69** [phenobarbital/asyncdb#1025]
Hi,

I'm using this repository in a private project of mine with `Python 3.11.6` and `Ubuntu 22.04`. 

Recently I decided to update my Python interpreter to the "new" version `3.12.2`. So I created a new environment, installed it with `pip` all requirements, including `asyncdb`.

But an error broke the requirements installation. After some time, I see that `asyncdb` raises the following error:

```bash
(my-env) andre@pc:~$ pip install asyncdb[odbc,msqlserver]
Collecting asyncdb[msqlserver,odbc]
  Using cached asyncdb-2.6.17-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)
Collecting numpy==1.24.2 (from asyncdb[msqlserver,odbc])
  Using cached numpy-1.24.2.tar.gz (10.9 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [33 lines of output]
      Traceback (most recent call last):
        File "/home/andre/.virtualenvs/my-env/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/home/andre/.virtualenvs/my-env/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/home/andre/.virtualenvs/my-env/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 112, in get_requires_for_build_wheel
          backend = _build_backend()
                    ^^^^^^^^^^^^^^^^
        File "/home/andre/.virtualenvs/my-env/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 77, in _build_backend
          obj = import_module(mod_path)
                ^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.12/importlib/__init__.py", line 90, in import_module
          return _bootstrap._gcd_import(name[level:], package, level)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
        File "<frozen importlib._bootstrap_external>", line 995, in exec_module
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "/tmp/pip-build-env-7mal0xc5/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 16, in <module>
          import setuptools.version
        File "/tmp/pip-build-env-7mal0xc5/overlay/lib/python3.12/site-packages/setuptools/version.py", line 1, in <module>
          import pkg_resources
        File "/tmp/pip-build-env-7mal0xc5/overlay/lib/python3.12/site-packages/pkg_resources/__init__.py", line 2172, in <module>
          register_finder(pkgutil.ImpImporter, find_on_path)
                          ^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

The error occurs because numpy has not been updated to the latest version. In fact, `Python3.12` removed `importlib.abc.Finder`, `pkgutil.ImpImporter` and `pkgutil.ImpLoader` ([as the documentation says](https://docs.python.org/3/whatsnew/3.12.html#importlib)).

Obviously with the well-known command `pip install asyncdb[odbc,msqlserver]` I get the same error.

Thanks in advance.

**Comment by phenobarbital** [comment]
nice catch!, the issue was not only on Numpy but also on pendulum (used by some datetime validations), I upgraded numpy and pendulum, and now Asyncdb can be built on Python 3.12 (version **2.6.18** Tested on: Python **3.12.2**), try to upgrade to last version.
_FYI_: I added Python 3.12.2 to the GitHub actions as well.

**Comment by AndreVale69** [comment]
Ok, now Numpy works!

But now I have a problem with `pyodbc`. Version `4.0.35` does not support Python 3.12. I get the following error after `pip install asyncdb[odbc,msqlserver]==2.6.18`:

```bash
Building wheels for collected packages: pyodbc
  Building wheel for pyodbc (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pyodbc (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [35 lines of output]
      running bdist_wheel
      running build
      running build_ext
      building 'pyodbc' extension
      creating build
      creating build/temp.linux-x86_64-cpython-312
      creating build/temp.linux-x86_64-cpython-312/src
      x86_64-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DPYODBC_VERSION=4.0.35 -I/home/andre/.virtualenvs/my-env/include -I/usr/include/python3.12 -c src/buffer.cpp -o build/temp.linux-x86_64-cpython-312/src/buffer.o -Wno-write-strings
      In file included from src/pyodbc.h:172,
                       from src/buffer.cpp:12:
      src/pyodbccompat.h: In function ‘PyObject* Text_New(Py_ssize_t)’:
      src/pyodbccompat.h:75:12: error: ‘PyUnicode_FromUnicode’ was not declared in this scope; did you mean ‘PyUnicode_FromString’?
         75 |     return PyUnicode_FromUnicode(0, length);
            |            ^~~~~~~~~~~~~~~~~~~~~
            |            PyUnicode_FromString
      src/pyodbccompat.h: In function ‘Py_UNICODE* Text_Buffer(PyObject*)’:
      src/pyodbccompat.h:86:12: error: ‘PyUnicode_AS_UNICODE’ was not declared in this scope; did you mean ‘PyUnicode_AsUCS4’?
         86 |     return PyUnicode_AS_UNICODE(o);
            |            ^~~~~~~~~~~~~~~~~~~~
            |            PyUnicode_AsUCS4
      src/pyodbccompat.h: In function ‘Py_ssize_t Text_Size(PyObject*)’:
      src/pyodbccompat.h:126:40: error: ‘PyUnicode_GET_SIZE’ was not declared in this scope; did you mean ‘PyDict_GET_SIZE’?
        126 |     return (o && PyUnicode_Check(o)) ? PyUnicode_GET_SIZE(o) : 0;
            |                                        ^~~~~~~~~~~~~~~~~~
            |                                        PyDict_GET_SIZE
      src/pyodbccompat.h: In function ‘Py_ssize_t TextCopyToUnicode(Py_UNICODE*, PyObject*)’:
      src/pyodbccompat.h:146:26: error: ‘PyUnicode_GET_SIZE’ was not declared in this scope; did you mean ‘PyDict_GET_SIZE’?
        146 |         Py_ssize_t cch = PyUnicode_GET_SIZE(o);
            |                          ^~~~~~~~~~~~~~~~~~
            |                          PyDict_GET_SIZE
      src/pyodbccompat.h:147:24: error: ‘PyUnicode_AS_UNICODE’ was not declared in this scope; did you mean ‘PyUnicode_AsUCS4’?
        147 |         memcpy(buffer, PyUnicode_AS_UNICODE(o), cch * sizeof(Py_UNICODE));
            |                        ^~~~~~~~~~~~~~~~~~~~
            |                        PyUnicode_AsUCS4
      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyodbc
Failed to build pyodbc
ERROR: Could not build wheels for pyodbc, which is required to install pyproject.toml-based projects
``` 

I analysed the error and noticed that the `pyodbc` team released the [version 5.0.0](https://github.com/mkleehammer/pyodbc/releases/tag/5.0.0) which supports Python 3.12 ([the pyodbc release list that I check](https://github.com/mkleehammer/pyodbc/releases)). So I think this is the problem.

To support my hypothesis, I created a Python 3.11 environment, installed the `asyncdb` package with `pip` (as above) and didn't get any errors (expected result, since the `4.0.35` `pyodbc` release supports Python 3.11, [according to the release documentation](https://github.com/mkleehammer/pyodbc/releases/tag/4.0.35)).

Thanks in advance for your fixes!

**Comment by phenobarbital** [comment]
Yes, there are several packages not compatible with python > 3.12 (memcache using pylibmc, oracledb, duckdb, pendulum), at this time, pylibmc requires compilation:
```
  Using cached pylibmc-1.6.3.tar.gz (63 kB)
  Installing build dependencies ... done
```
 ascylla (used by ScyllaDB connector) is only working on python 3.11.
 
 I released version 2.6.20 to fix those issues (I built over python 3.12.2).

**Comment by AndreVale69** [comment]
This issue can be closed for me. Now I don't find any error when I run `pip install asyncdb[odbc,msqlserver]`. Thanks for your help!
